src/bulk_insert/bulk_insert.c

```diff
@@ -7,388 +7,211 @@
-            return NULL;
-        }
-        *argc -= label_count * 2;
-
-        long long edge_count;
-        for(int i = 0; i < label_count; i++) {
-            labelRelations[i].label = RedisModule_StringPtrLen(*argv++, NULL);
-            if(RedisModule_StringToLongLong(*argv++, &labelRelations[i].edge_count) != REDISMODULE_OK) {
-                char *err;
-                asprintf(&err, "Bulk insert format error, failed to parse number of relations of type '%s'.", labelRelations[i].label);
-                _BulkInsert_Reply_With_Syntax_Error(ctx, err);
-                return NULL;
-            }
-            total_edges += labelRelations[i].edge_count;
-            LabelStore *s = GraphContext_GetStore(gc, labelRelations[i].label, STORE_EDGE);
-            if(!s) {
-                s = GraphContext_AddRelationType(gc, labelRelations[i].label);
-                // TODO: Add support for parsing properties on relations
-            }
-            labelRelations[i].label_id = s->id;
-        }
+    if (argc < 1) {
+        RedisModule_ReplyWithError(ctx, "Bulk insert format error, failed to parse bulk insert sections.");
+        return BULK_FAIL;
     }
 
-    if(*argc < relations_count * 2) {
-        _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to parse relations.");
-        return NULL;
-    }
-    *argc -= relations_count * 2;
-
-    // Introduce relations.
-    NodeID srcId;
-    NodeID destId;
-    for(int i = 0; i < label_count; i++) {
-        LabelRelation labelRelation = labelRelations[i];
-
-        for(int j = 0; j < labelRelation.edge_count; j++) {
-            if(RedisModule_StringToLongLong(*argv++, (long long*)&srcId) != REDISMODULE_OK) {
-                _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to read relation source node id.");
-                return NULL;
-            }
-            if(RedisModule_StringToLongLong(*argv++, (long long*)&destId) != REDISMODULE_OK) {
-                _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to read relation destination node id.");
-                return NULL;
-            }
-
-            Edge e;
-            Graph_ConnectNodes(gc->g, srcId, destId, labelRelation.label_id, &e);
-        }
+    // Read the number of node tokens
+    long long node_token_count;
+    long long relation_token_count;
+    if (RedisModule_StringToLongLong(*argv++, &node_token_count)  != REDISMODULE_OK) {
+        RedisModule_ReplyWithError(ctx, "Error parsing number of node descriptor tokens.");
+        return BULK_FAIL;
     }
-    return argv;
-}
-
-int BulkInsert(RedisModuleCtx *ctx, GraphContext *gc, size_t *nodes, size_t *edges, RedisModuleString **argv, int argc) {
 
-    if(argc < 1) {
-        _BulkInsert_Reply_With_Syntax_Error(ctx, "Bulk insert format error, failed to parse bulk insert sections.");
+    if (RedisModule_StringToLongLong(*argv++, &relation_token_count)  != REDISMODULE_OK) {
```

Comment:
Are we guaranteed to have at least 2 arguments, the check above verifies that we have at-least one, Also whenever we read from argv we should update argc

Reply:
Agreed on the conditional change. argc gets reduced by 2 after these two reads, I don't feel that multiple calls are necessary.